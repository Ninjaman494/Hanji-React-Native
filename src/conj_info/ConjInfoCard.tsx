import React from "react";
import { Conjugation } from "../hooks/useConjugations";
import { Badge, Text, Title, useTheme } from "react-native-paper";
import { View } from "react-native";
import BaseCard from "../components/BaseCard";

export type ConjInfoCardProps = {
  conjugation: Conjugation;
};

const ConjInfoCard: React.FC<ConjInfoCardProps> = ({ conjugation }) => {
  const { colors, padding } = useTheme();

  const styles = {
    card: {
      marginHorizontal: padding?.horizontal,
    },
    badge: {
      paddingHorizontal: padding?.horizontal,
      marginRight: padding?.horizontal,
      backgroundColor: colors?.accent,
    },
    title: {
      fontSize: 18,
      color: colors?.primary,
      marginTop: 16,
      marginBottom: padding?.vertical,
    },
    conjugation: {
      fontSize: 24,
      marginTop: -16,
    },
    text: {
      fontSize: 20,
    },
    secondary: {
      fontSize: 18,
      color: "#808080",
    },
  };

  const reasons = conjugation?.reasons.map((reason) => {
    const index = reason.indexOf("(");
    return {
      type: reason.substring(0, index),
      details: reason.substring(index),
    };
  });

  return (
    <BaseCard
      title={conjugation?.name}
      style={styles.card}
      rightIcon={() => (
        <Badge size={30} visible={conjugation?.honorific} style={styles.badge}>
          Honorific
        </Badge>
      )}
    >
      <View style={{ paddingHorizontal: padding?.horizontal }}>
        <Text style={styles.conjugation}>{conjugation?.conjugation}</Text>
        <Title style={styles.title}>Pronunciation</Title>
        <Text style={styles.text}>{conjugation?.pronunciation}</Text>
        <Text style={styles.secondary}>{conjugation?.romanization}</Text>
        <Title style={styles.title}>Explanation</Title>
        {reasons.map((reason, index) => (
          <View style={{ marginBottom: padding?.vertical }} key={index}>
            <Text style={styles.text}>{reason.type}</Text>
            <Text style={styles.secondary}>{reason.details}</Text>
          </View>
        ))}
      </View>
    </BaseCard>
  );
};

export default ConjInfoCard;
