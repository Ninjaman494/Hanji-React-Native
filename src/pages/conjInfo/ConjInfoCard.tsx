import { BaseCard, HonorificBadge } from "components";
import { Conjugation } from "hooks/useConjugations";
import React from "react";
import { StyleSheet, View } from "react-native";
import { Text, Title, useTheme } from "react-native-paper";
import toTitleCase from "utils/toTitleCase";

export type ConjInfoCardProps = {
  conjugation: Conjugation;
};

const ConjInfoCard: React.FC<ConjInfoCardProps> = ({ conjugation }) => {
  const { colors, padding, textSizes } = useTheme();

  const styles = StyleSheet.create({
    card: {
      marginHorizontal: padding?.horizontal,
      marginBottom: 8,
    },
    title: {
      fontSize: textSizes?.cardTitle,
      fontWeight: "400",
      color: colors?.primary,
      marginTop: 16,
      marginBottom: padding?.vertical,
    },
    conjugation: {
      fontSize: 24,
      marginTop: -24,
    },
    text: {
      fontSize: 20,
    },
    secondary: {
      fontSize: 18,
      color: "#808080",
    },
  });

  const reasons = conjugation?.reasons.map((reason) => {
    const index = reason.indexOf("(");
    return {
      type: reason.substring(0, index),
      details: reason.substring(index),
    };
  });

  return (
    <BaseCard
      title={toTitleCase(conjugation?.name)}
      style={styles.card}
      rightIcon={() => <HonorificBadge visible={conjugation?.honorific} />}
    >
      <View style={{ paddingHorizontal: padding?.horizontal }}>
        <Text style={styles.conjugation} selectable>
          {conjugation?.conjugation}
        </Text>
        <Title style={styles.title}>Pronunciation</Title>
        <Text style={styles.text} selectable>
          {conjugation?.pronunciation}
        </Text>
        <Text style={styles.secondary} selectable>
          {conjugation?.romanization}
        </Text>
        <Title style={styles.title}>Explanation</Title>
        {reasons.map((reason, index) => (
          <View style={{ marginBottom: padding?.vertical }} key={index}>
            <Text style={styles.text} selectable>
              {reason.type}
            </Text>
            <Text style={styles.secondary} selectable>
              {reason.details}
            </Text>
          </View>
        ))}
      </View>
    </BaseCard>
  );
};

export default ConjInfoCard;
