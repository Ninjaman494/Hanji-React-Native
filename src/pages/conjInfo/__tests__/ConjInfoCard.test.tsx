import React from "react";
import { render } from "@testing-library/react-native";
import ConjInfoCard from "../ConjInfoCard";
import {
  ConjugationName,
  ConjugationType,
  Formality,
  Tense,
} from "utils/conjugationTypes";

describe("ConjInfoCard component", () => {
  const conjugation = {
    name: "name" as ConjugationName,
    conjugation: "conj",
    type: ConjugationType.DECLARATIVE_PRESENT,
    tense: Tense.PRESENT,
    speechLevel: Formality.INFORMAL_LOW,
    honorific: false,
    pronunciation: "pronunciation",
    romanization: "romanization",
    reasons: ["reason 1 (details)", "reason 2 (details)", "reason 3 (details)"],
  };

  it("displays a regular conjugation", () => {
    const component = render(<ConjInfoCard conjugation={conjugation} />);

    expect(component).toMatchSnapshot();
  });

  it("displays an honorific conjugation", () => {
    const component = render(
      <ConjInfoCard conjugation={{ ...conjugation, honorific: true }} />
    );

    expect(component).toMatchSnapshot();
  });
});
