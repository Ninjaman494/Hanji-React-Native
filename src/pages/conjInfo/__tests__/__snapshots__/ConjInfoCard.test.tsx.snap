// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ConjInfoCard component displays a regular conjugation 1`] = `
<View
  collapsable={false}
  pointerEvents="box-none"
  style={
    {
      "flex": 1,
    }
  }
>
  <View
    collapsable={false}
    nativeID="animatedComponent"
    style={
      {
        "backgroundColor": "#ffffff",
        "borderColor": "rgba(0, 0, 0, 0.12)",
        "borderRadius": 10,
        "elevation": 5,
        "shadowColor": "#000000",
        "shadowOffset": {
          "height": 4,
          "width": 0,
        },
        "shadowOpacity": 0.24,
        "shadowRadius": 5,
      }
    }
  >
    <View
      accessibilityState={
        {
          "disabled": true,
        }
      }
      accessible={true}
      focusable={false}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        {
          "flexGrow": 1,
          "flexShrink": 1,
        }
      }
    >
      <View
        style={
          [
            {
              "alignItems": "center",
              "flexDirection": "row",
              "justifyContent": "space-between",
              "paddingLeft": 16,
            },
            {
              "minHeight": 72,
            },
            {
              "marginTop": -8,
            },
          ]
        }
      >
        <View
          style={
            [
              {
                "flex": 1,
                "flexDirection": "column",
                "justifyContent": "center",
              },
            ]
          }
        >
          <Text
            numberOfLines={1}
            style={
              [
                {
                  "color": "#000000",
                  "fontFamily": "System",
                  "fontWeight": "400",
                },
                {
                  "textAlign": "left",
                },
                [
                  {
                    "textAlign": "left",
                  },
                  {
                    "color": "rgba(0, 0, 0, 0.87)",
                    "fontFamily": "System",
                    "fontWeight": "500",
                    "writingDirection": "ltr",
                  },
                  [
                    {
                      "fontSize": 20,
                      "letterSpacing": 0.15,
                      "lineHeight": 30,
                      "marginVertical": 2,
                    },
                    [
                      {
                        "minHeight": 30,
                      },
                      {
                        "marginBottom": 2,
                      },
                      {
                        "color": "#3F51B5",
                        "fontSize": 16,
                        "fontWeight": "400",
                      },
                    ],
                  ],
                ],
              ]
            }
          >
            Name
          </Text>
        </View>
        <View>
          <Text
            collapsable={false}
            nativeID="animatedComponent"
            numberOfLines={1}
            style={
              {
                "alignSelf": "flex-end",
                "backgroundColor": "#F44336",
                "borderRadius": 15,
                "color": "#ffffff",
                "fontFamily": "System",
                "fontSize": 15,
                "fontWeight": "500",
                "height": 30,
                "lineHeight": 30,
                "marginRight": 16,
                "minWidth": 30,
                "opacity": 0,
                "overflow": "hidden",
                "paddingHorizontal": 16,
                "textAlign": "center",
                "textAlignVertical": "center",
              }
            }
          >
            Honorific
          </Text>
        </View>
      </View>
      <View
        style={
          [
            {
              "paddingHorizontal": 16,
            },
            undefined,
            {
              "paddingBottom": 8,
              "paddingHorizontal": 0,
            },
          ]
        }
      >
        <View
          style={
            {
              "paddingHorizontal": 16,
            }
          }
        >
          <Text
            selectable={true}
            style={
              [
                {
                  "color": "#000000",
                  "fontFamily": "System",
                  "fontWeight": "400",
                },
                {
                  "textAlign": "left",
                },
                {
                  "fontSize": 24,
                  "marginTop": -24,
                },
              ]
            }
          >
            conj
          </Text>
          <Text
            style={
              [
                {
                  "color": "#000000",
                  "fontFamily": "System",
                  "fontWeight": "400",
                },
                {
                  "textAlign": "left",
                },
                [
                  {
                    "textAlign": "left",
                  },
                  {
                    "color": "rgba(0, 0, 0, 0.87)",
                    "fontFamily": "System",
                    "fontWeight": "500",
                    "writingDirection": "ltr",
                  },
                  [
                    {
                      "fontSize": 20,
                      "letterSpacing": 0.15,
                      "lineHeight": 30,
                      "marginVertical": 2,
                    },
                    {
                      "color": "#3F51B5",
                      "fontSize": 16,
                      "fontWeight": "400",
                      "marginBottom": 8,
                      "marginTop": 16,
                    },
                  ],
                ],
              ]
            }
          >
            Pronunciation
          </Text>
          <Text
            selectable={true}
            style={
              [
                {
                  "color": "#000000",
                  "fontFamily": "System",
                  "fontWeight": "400",
                },
                {
                  "textAlign": "left",
                },
                {
                  "fontSize": 20,
                },
              ]
            }
          >
            pronunciation
          </Text>
          <Text
            selectable={true}
            style={
              [
                {
                  "color": "#000000",
                  "fontFamily": "System",
                  "fontWeight": "400",
                },
                {
                  "textAlign": "left",
                },
                {
                  "color": "#808080",
                  "fontSize": 18,
                },
              ]
            }
          >
            romanization
          </Text>
          <Text
            style={
              [
                {
                  "color": "#000000",
                  "fontFamily": "System",
                  "fontWeight": "400",
                },
                {
                  "textAlign": "left",
                },
                [
                  {
                    "textAlign": "left",
                  },
                  {
                    "color": "rgba(0, 0, 0, 0.87)",
                    "fontFamily": "System",
                    "fontWeight": "500",
                    "writingDirection": "ltr",
                  },
                  [
                    {
                      "fontSize": 20,
                      "letterSpacing": 0.15,
                      "lineHeight": 30,
                      "marginVertical": 2,
                    },
                    {
                      "color": "#3F51B5",
                      "fontSize": 16,
                      "fontWeight": "400",
                      "marginBottom": 8,
                      "marginTop": 16,
                    },
                  ],
                ],
              ]
            }
          >
            Explanation
          </Text>
          <View
            style={
              {
                "marginBottom": 8,
              }
            }
          >
            <Text
              selectable={true}
              style={
                [
                  {
                    "color": "#000000",
                    "fontFamily": "System",
                    "fontWeight": "400",
                  },
                  {
                    "textAlign": "left",
                  },
                  {
                    "fontSize": 20,
                  },
                ]
              }
            >
              reason 1 
            </Text>
            <Text
              selectable={true}
              style={
                [
                  {
                    "color": "#000000",
                    "fontFamily": "System",
                    "fontWeight": "400",
                  },
                  {
                    "textAlign": "left",
                  },
                  {
                    "color": "#808080",
                    "fontSize": 18,
                  },
                ]
              }
            >
              (details)
            </Text>
          </View>
          <View
            style={
              {
                "marginBottom": 8,
              }
            }
          >
            <Text
              selectable={true}
              style={
                [
                  {
                    "color": "#000000",
                    "fontFamily": "System",
                    "fontWeight": "400",
                  },
                  {
                    "textAlign": "left",
                  },
                  {
                    "fontSize": 20,
                  },
                ]
              }
            >
              reason 2 
            </Text>
            <Text
              selectable={true}
              style={
                [
                  {
                    "color": "#000000",
                    "fontFamily": "System",
                    "fontWeight": "400",
                  },
                  {
                    "textAlign": "left",
                  },
                  {
                    "color": "#808080",
                    "fontSize": 18,
                  },
                ]
              }
            >
              (details)
            </Text>
          </View>
          <View
            style={
              {
                "marginBottom": 8,
              }
            }
          >
            <Text
              selectable={true}
              style={
                [
                  {
                    "color": "#000000",
                    "fontFamily": "System",
                    "fontWeight": "400",
                  },
                  {
                    "textAlign": "left",
                  },
                  {
                    "fontSize": 20,
                  },
                ]
              }
            >
              reason 3 
            </Text>
            <Text
              selectable={true}
              style={
                [
                  {
                    "color": "#000000",
                    "fontFamily": "System",
                    "fontWeight": "400",
                  },
                  {
                    "textAlign": "left",
                  },
                  {
                    "color": "#808080",
                    "fontSize": 18,
                  },
                ]
              }
            >
              (details)
            </Text>
          </View>
        </View>
      </View>
    </View>
  </View>
</View>
`;

exports[`ConjInfoCard component displays an honorific conjugation 1`] = `
<View
  collapsable={false}
  pointerEvents="box-none"
  style={
    {
      "flex": 1,
    }
  }
>
  <View
    collapsable={false}
    nativeID="animatedComponent"
    style={
      {
        "backgroundColor": "#ffffff",
        "borderColor": "rgba(0, 0, 0, 0.12)",
        "borderRadius": 10,
        "elevation": 5,
        "shadowColor": "#000000",
        "shadowOffset": {
          "height": 4,
          "width": 0,
        },
        "shadowOpacity": 0.24,
        "shadowRadius": 5,
      }
    }
  >
    <View
      accessibilityState={
        {
          "disabled": true,
        }
      }
      accessible={true}
      focusable={false}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        {
          "flexGrow": 1,
          "flexShrink": 1,
        }
      }
    >
      <View
        style={
          [
            {
              "alignItems": "center",
              "flexDirection": "row",
              "justifyContent": "space-between",
              "paddingLeft": 16,
            },
            {
              "minHeight": 72,
            },
            {
              "marginTop": -8,
            },
          ]
        }
      >
        <View
          style={
            [
              {
                "flex": 1,
                "flexDirection": "column",
                "justifyContent": "center",
              },
            ]
          }
        >
          <Text
            numberOfLines={1}
            style={
              [
                {
                  "color": "#000000",
                  "fontFamily": "System",
                  "fontWeight": "400",
                },
                {
                  "textAlign": "left",
                },
                [
                  {
                    "textAlign": "left",
                  },
                  {
                    "color": "rgba(0, 0, 0, 0.87)",
                    "fontFamily": "System",
                    "fontWeight": "500",
                    "writingDirection": "ltr",
                  },
                  [
                    {
                      "fontSize": 20,
                      "letterSpacing": 0.15,
                      "lineHeight": 30,
                      "marginVertical": 2,
                    },
                    [
                      {
                        "minHeight": 30,
                      },
                      {
                        "marginBottom": 2,
                      },
                      {
                        "color": "#3F51B5",
                        "fontSize": 16,
                        "fontWeight": "400",
                      },
                    ],
                  ],
                ],
              ]
            }
          >
            Name
          </Text>
        </View>
        <View>
          <Text
            collapsable={false}
            nativeID="animatedComponent"
            numberOfLines={1}
            style={
              {
                "alignSelf": "flex-end",
                "backgroundColor": "#F44336",
                "borderRadius": 15,
                "color": "#ffffff",
                "fontFamily": "System",
                "fontSize": 15,
                "fontWeight": "500",
                "height": 30,
                "lineHeight": 30,
                "marginRight": 16,
                "minWidth": 30,
                "opacity": 1,
                "overflow": "hidden",
                "paddingHorizontal": 16,
                "textAlign": "center",
                "textAlignVertical": "center",
              }
            }
          >
            Honorific
          </Text>
        </View>
      </View>
      <View
        style={
          [
            {
              "paddingHorizontal": 16,
            },
            undefined,
            {
              "paddingBottom": 8,
              "paddingHorizontal": 0,
            },
          ]
        }
      >
        <View
          style={
            {
              "paddingHorizontal": 16,
            }
          }
        >
          <Text
            selectable={true}
            style={
              [
                {
                  "color": "#000000",
                  "fontFamily": "System",
                  "fontWeight": "400",
                },
                {
                  "textAlign": "left",
                },
                {
                  "fontSize": 24,
                  "marginTop": -24,
                },
              ]
            }
          >
            conj
          </Text>
          <Text
            style={
              [
                {
                  "color": "#000000",
                  "fontFamily": "System",
                  "fontWeight": "400",
                },
                {
                  "textAlign": "left",
                },
                [
                  {
                    "textAlign": "left",
                  },
                  {
                    "color": "rgba(0, 0, 0, 0.87)",
                    "fontFamily": "System",
                    "fontWeight": "500",
                    "writingDirection": "ltr",
                  },
                  [
                    {
                      "fontSize": 20,
                      "letterSpacing": 0.15,
                      "lineHeight": 30,
                      "marginVertical": 2,
                    },
                    {
                      "color": "#3F51B5",
                      "fontSize": 16,
                      "fontWeight": "400",
                      "marginBottom": 8,
                      "marginTop": 16,
                    },
                  ],
                ],
              ]
            }
          >
            Pronunciation
          </Text>
          <Text
            selectable={true}
            style={
              [
                {
                  "color": "#000000",
                  "fontFamily": "System",
                  "fontWeight": "400",
                },
                {
                  "textAlign": "left",
                },
                {
                  "fontSize": 20,
                },
              ]
            }
          >
            pronunciation
          </Text>
          <Text
            selectable={true}
            style={
              [
                {
                  "color": "#000000",
                  "fontFamily": "System",
                  "fontWeight": "400",
                },
                {
                  "textAlign": "left",
                },
                {
                  "color": "#808080",
                  "fontSize": 18,
                },
              ]
            }
          >
            romanization
          </Text>
          <Text
            style={
              [
                {
                  "color": "#000000",
                  "fontFamily": "System",
                  "fontWeight": "400",
                },
                {
                  "textAlign": "left",
                },
                [
                  {
                    "textAlign": "left",
                  },
                  {
                    "color": "rgba(0, 0, 0, 0.87)",
                    "fontFamily": "System",
                    "fontWeight": "500",
                    "writingDirection": "ltr",
                  },
                  [
                    {
                      "fontSize": 20,
                      "letterSpacing": 0.15,
                      "lineHeight": 30,
                      "marginVertical": 2,
                    },
                    {
                      "color": "#3F51B5",
                      "fontSize": 16,
                      "fontWeight": "400",
                      "marginBottom": 8,
                      "marginTop": 16,
                    },
                  ],
                ],
              ]
            }
          >
            Explanation
          </Text>
          <View
            style={
              {
                "marginBottom": 8,
              }
            }
          >
            <Text
              selectable={true}
              style={
                [
                  {
                    "color": "#000000",
                    "fontFamily": "System",
                    "fontWeight": "400",
                  },
                  {
                    "textAlign": "left",
                  },
                  {
                    "fontSize": 20,
                  },
                ]
              }
            >
              reason 1 
            </Text>
            <Text
              selectable={true}
              style={
                [
                  {
                    "color": "#000000",
                    "fontFamily": "System",
                    "fontWeight": "400",
                  },
                  {
                    "textAlign": "left",
                  },
                  {
                    "color": "#808080",
                    "fontSize": 18,
                  },
                ]
              }
            >
              (details)
            </Text>
          </View>
          <View
            style={
              {
                "marginBottom": 8,
              }
            }
          >
            <Text
              selectable={true}
              style={
                [
                  {
                    "color": "#000000",
                    "fontFamily": "System",
                    "fontWeight": "400",
                  },
                  {
                    "textAlign": "left",
                  },
                  {
                    "fontSize": 20,
                  },
                ]
              }
            >
              reason 2 
            </Text>
            <Text
              selectable={true}
              style={
                [
                  {
                    "color": "#000000",
                    "fontFamily": "System",
                    "fontWeight": "400",
                  },
                  {
                    "textAlign": "left",
                  },
                  {
                    "color": "#808080",
                    "fontSize": 18,
                  },
                ]
              }
            >
              (details)
            </Text>
          </View>
          <View
            style={
              {
                "marginBottom": 8,
              }
            }
          >
            <Text
              selectable={true}
              style={
                [
                  {
                    "color": "#000000",
                    "fontFamily": "System",
                    "fontWeight": "400",
                  },
                  {
                    "textAlign": "left",
                  },
                  {
                    "fontSize": 20,
                  },
                ]
              }
            >
              reason 3 
            </Text>
            <Text
              selectable={true}
              style={
                [
                  {
                    "color": "#000000",
                    "fontFamily": "System",
                    "fontWeight": "400",
                  },
                  {
                    "textAlign": "left",
                  },
                  {
                    "color": "#808080",
                    "fontSize": 18,
                  },
                ]
              }
            >
              (details)
            </Text>
          </View>
        </View>
      </View>
    </View>
  </View>
</View>
`;
